Allow distributed breakpoint to skip the first few calls

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129511

Summary:
PDB allows to do conditional breakpoint but the ability won't work in the distributed environment. We can still do conditional breakpoint by doing the following:

```
counter = 0

global counter
count += 1
if counter > 100:
  dist.breakpoint()
```

This PR makes dist.breakpoint() support this feature as a syntax sugar.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l @yf225 @chauhang

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

wconstab commented on torch/distributed/__init__.py:
```diff
@@ -73,14 +75,25 @@ def interaction(self, *args, **kwargs):
+            counter = _breakpoint_cache.get(key, 0) + 1
+            _breakpoint_cache[key] = counter
+            if counter <= skip:
+                print(f"Skip the breakpoint, {counter=}.")
```
logger.debug? or .info maybe? but probably not print?

wconstab reviewed (approved):
lgtm with one nit for printing. thanks for adding this!

fegin committed:
Update

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update on "Allow distributed breakpoint to skip the first few calls"


Summary:
PDB allows to do conditional breakpoint but the ability won't work in the distributed environment. We can still do conditional breakpoint by doing the following:

```
counter = 0

global counter
count += 1
if counter > 100:
  dist.breakpoint()
```

This PR makes dist.breakpoint() support this feature as a syntax sugar.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
